internal/runtime/client/client_test.go

```diff
@@ -0,0 +1,169 @@
+		},
+		{
+			name:     "should error if hooks is not registered with catalog",
+			request:  &fakev1alpha1.FakeRequest{},
+			response: &fakev1alpha1.FakeResponse{},
+			opts: &httpCallOptions{
+				catalog: catalog.New(),
+			},
+			wantErr: true,
+		},
+		{
+			name: "should succeed for valid request and response objects",
+			request: &fakev1alpha1.FakeRequest{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "FakeRequest",
+					APIVersion: fakev1alpha1.GroupVersion.Identifier(),
+				},
+			},
+			response: &fakev1alpha1.FakeResponse{},
+			opts: func() *httpCallOptions {
+				c := catalog.New()
+				c.AddHook(
+					fakev1alpha1.GroupVersion,
+					fakev1alpha1.FakeHook,
+					&catalog.HookMeta{},
+				)
+
+				// get same gvh for hook by using the FakeHook and catalog
+				gvh, err := c.GroupVersionHook(fakev1alpha1.FakeHook)
+				g.Expect(err).To(Succeed())
+
+				return &httpCallOptions{
+					catalog: c,
+					gvh:     gvh,
+				}
+			}(),
+			wantErr: false,
+		},
+		{
+			name: "should success if response and response are valid objects - with conversion",
+			request: &fakev1alpha2.FakeRequest{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "FakeRequest",
+					APIVersion: fakev1alpha2.GroupVersion.Identifier(),
+				},
+			},
+			response: &fakev1alpha2.FakeResponse{},
+			opts: func() *httpCallOptions {
+				c := catalog.New()
+				// register fakev1alpha1 to enable conversion
+				g.Expect(fakev1alpha1.AddToCatalog(c)).To(Succeed())
+				c.AddHook(
+					fakev1alpha1.GroupVersion,
+					fakev1alpha1.FakeHook,
+					&catalog.HookMeta{},
+				)
+
+				// get same gvh for hook by using the FakeHook and catalog
+				gvh, err := c.GroupVersionHook(fakev1alpha1.FakeHook)
+				g.Expect(err).To(Succeed())
+
+				return &httpCallOptions{
+					catalog: c,
+					gvh:     gvh,
+				}
+			}(),
+			wantErr: false,
+		},
+	}
+	for _, tt := range tableTests {
+		t.Run(tt.name, func(t *testing.T) {
+			// a http server is only required if we have a valid catalog, otherwise httpCall will not reach out to the server
+			if tt.opts != nil && tt.opts.catalog != nil {
+				// create http server with fakeHookHandler
+				mux := http.NewServeMux()
+				mux.HandleFunc("/", fakeHookHandler)
+				srv := httptest.NewServer(mux)
+				defer srv.Close()
+
+				// set url to srv for in tt.opts
+				tt.opts.config.URL = pointer.String(srv.URL)
+			}
+
+			assert := g.Expect(httpCall(context.TODO(), tt.request, tt.response, tt.opts))
+			if tt.wantErr {
+				assert.To(HaveOccurred())
+			} else {
+				assert.To(Succeed())
+			}
+		})
+	}
+}
+
+func fakeHookHandler(w http.ResponseWriter, r *http.Request) {
+	response := &fakev1alpha1.FakeResponse{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "FakeHookResponse",
+			APIVersion: fakev1alpha1.GroupVersion.Identifier(),
+		},
+		Second: "",
+		First:  1,
+	}
+	respBody, err := json.Marshal(response)
+	if err != nil {
+		panic(err)
+	}
+	w.WriteHeader(http.StatusOK)
+	_, _ = w.Write(respBody)
+}
```

Comment:
I know it's WIP, but just a suggestion already. Table test for urlForExtension would be really nice